src/Microsoft.AspNetCore.Sockets.Common/TextMessageBatchFormatter.cs

```diff
@@ -0,0 +1,239 @@
+            if(message.Type == MessageType.Binary)
+            {
+                var payload = Encoding.UTF8.GetBytes(Convert.ToBase64String(message.Payload.Buffer.ToArray()));
+                payload.CopyTo(target.Slice(cursor, payload.Length));
+                cursor += payload.Length;
+            } else
+            {
+                message.Payload.Buffer.CopyTo(target.Slice(cursor, message.Payload.Buffer.Length));
+                cursor += message.Payload.Buffer.Length;
+            }
+
+            // Terminator
+            target[cursor] = (byte)';';
+            cursor++;
+        }
+
+        public static int MeasureMessages(IReadOnlyList<Message> messages)
+        {
+            var size = 1; // 1 character for the 'T'
+            foreach(var message in messages)
+            {
+                size += MeasureMessage(message);
+            }
+            return size;
+        }
+
+        private static int MeasureMessage(Message message)
+        {
+            var size = 4; // two ':', one ';' and the type flag
+
+            // Add enough characters for the payload
+            var len = message.Type == MessageType.Binary ?
+                // Length is measured as Base-64 character length
+                (int)(4 * Math.Ceiling(((double)message.Payload.Buffer.Length / 3))) :
+                // Length is just number of bytes in the (already UTF-8 encoded) payload
+                message.Payload.Buffer.Length;
+            size += len;
+
+            // Add enough character to represent the length in digits
+            // Log10(0) == Infinity, so don't allow that to happen.
+            var digits = len == 0 ? 1 : (int)Math.Floor(Math.Log10(len)) + 1;
+            size += digits;
+            return size;
+        }
+
+        private static PreservedBuffer ParsePayload(ReadOnlySpan<byte> data, int length, MessageType messageType, ref int cursor)
+        {
+            int start = cursor;
+
+            // We know exactly where the end is. The last byte is cursor + length
+            cursor += length;
+
+            // Verify the length and trailer
+            if (cursor >= data.Length)
+            {
+                throw new FormatException("Unexpected end-of-message while reading Payload field.");
+            }
+            if (data[cursor] != ';')
+            {
+                throw new FormatException("Payload is missing trailer character ';'.");
+            }
+
+            // Read the data into a buffer
+            var buffer = new byte[length];
+            data.Slice(start, length).CopyTo(buffer);
+
+            // If the message is binary, we need to convert from Base64
+            if (messageType == MessageType.Binary)
+            {
+                // TODO: Use System.Binary.Base64 to handle this with less allocation
+
+                // Parse the data as Base64
+                var str = Encoding.UTF8.GetString(buffer);
+                buffer = Convert.FromBase64String(str);
+            }
```

Comment:
You should not be copying out of data. You should use PrimitiveBarser to parse directly, and Base64 (decoder) to decode directly from the span.